pkg/api/validation/validation.go

```diff
@@ -1708,10 +1708,30 @@ func validateTolerations(tolerations []api.Toleration, fldPath *field.Path) fiel
 	for i, toleration := range tolerations {
 		idxPath := fldPath.Index(i)
 		// validate the toleration key
-		allErrors = append(allErrors, unversionedvalidation.ValidateLabelName(toleration.Key, idxPath.Child("key"))...)
+		if len(toleration.Key) > 0 {
+			allErrors = append(allErrors, unversionedvalidation.ValidateLabelName(toleration.Key, idxPath.Child("key"))...)
+		}
+
+		// empty toleration key with Exists operator and empty value means match all taints
+		if len(toleration.Key) == 0 && toleration.Operator != api.TolerationOpExists {
+			allErrors = append(allErrors, field.Invalid(idxPath.Child("operator"), toleration.Operator,
+				"operator must be Exists when `key` is empty, which means \"match all values and all keys\""))
+		}
+
+		if toleration.TolerationSeconds != nil {
+			if *toleration.TolerationSeconds <= 0 {
```

Comment:
`time.Duration` can be negative as well, so this check won't simply go away. The only doubt I have is whether we need to do this check at all. I don't know what our validation policies are - do we reject only things that would break the system in some way (e.g. when we're dividing by given value, or sth.), or we're trying hard to prevent user from doing stupid things. In this case passing negative toleration time wouldn't break anything - it'll be effectively treated as 0 by the system.